16942: ttnn.experimental.scatter_

jbbieniekTT opened the pull request:
### Ticket
#16942 

### Problem description
Provide `ttnn.experimental.scatter_` to work for input tensors of the size required by the client. For now, long input and index tensors cause excessive execution time, and this is to be covered by the issue #22926.

### What's changed
- Implemented `ttnn.experimental.scatter_`.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

jbbieniekTT committed:
16942: ttnn.experimental.scatter_

bbradelTT reviewed (commented):
We're all really pressed for time. Please try to have your PRs in a finished state before sending them out. For getting a first opinion, consider a draft PR and request a preliminary review from one person.

My main feedback:
- Don't have ???
- Have issue numbers next to TODOs
- Remove commented out code
- Remove dprints unless they're within a debug block
Also
- tile_guts_gen seems strange
- I can't tell what "memory_config testing!!!!!!!!!!" means. My best guess is that you're trying to remind yourself to add tests

jbbieniekTT commented:
@bbradelTT Hey, I apologise, I should have done as you said. I actually forgot to put it into draft state. I am working on cleaning all the stuff.

jbbieniekTT committed:
16942: fix the bug with Ht > 32

fix the titular bug

jbbieniekTT committed:
16942: remove tile_guts_gen

remove tile_guts_gen

jbbieniekTT committed:
16942: add multi-core and fix licenses

- replace single-core with multi-core for ttnn.experimental.scatter_
- fix licenses in source files

jbbieniekTT commented on ttnn/cpp/ttnn/operations/experimental/scatter/device/kernels/dataflow/reader_scatter.cpp:
```diff
@@ -0,0 +1,136 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
```
Done

jbbieniekTT committed:
16942: support for dtypes

support for any dtypes ready as soon as
ttnn.transpose works with integers

jbbieniekTT committed:
16942: provide the O(n^2) algorithm

Use a brute-force algorithm for `scatter`
to handle all shape cases

jbbieniekTT committed:
16942: remove division and modulo

get rid of titular arithmetic from kernels

jbbieniekTT committed:
16942: fix index bug

fix index bug

jbbieniekTT committed:
16942: fix bfloat16 and 1D cases

fix the titular cases

bbradelTT commented on ttnn/cpp/ttnn/operations/experimental/scatter/device/kernels/scatter_common.hpp:
```diff
@@ -0,0 +1,145 @@
+        get_compile_time_arg_val(15),
+        get_compile_time_arg_val(16),
+        get_compile_time_arg_val(17),
+        get_compile_time_arg_val(18)};
```
Would this benefit in any way from https://github.com/tenstorrent/tt-metal/pull/18379 ?

bbradelTT replied:
I looked into this. Doesn't seem to be a good way to not enumerate through all arg vals.

jbbieniekTT committed:
16942: address PR comments

jbbieniekTT committed:
16942: finish python doc for scatter

finish the doc in the scatter_pybind.cpp for
ttnn.experimental.scatter_ (example usage)

jbbieniekTT committed:
16942: standarize scatter's build

- provide a CMakeLists.txt for scatter
- modify ttnn/CMakeLists.txt

jbbieniekTT committed:
16942: pydoc for scatter

finish pydoc for scatter

jbbieniekTT committed:
16942: little fix

little fix for 1D reshaping

jbbieniekTT committed:
16942: small fix

small fix in the test

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/scatter/device/kernels/scatter_common.hpp:
```diff
@@ -0,0 +1,142 @@
+
+FORCE_INLINE uint32_t
+get_width_scalar_index(const uint32_t& tile_id, const uint32_t& face_x, const uint32_t& scalar_x) {
+    return tile_id * tt::constants::TILE_WIDTH + face_x * (tt::constants::TILE_WIDTH >> 1) + scalar_x;
```
tile height and width and not constants since ~6 months ago (the feature was called tiny tiles and was added to improve perf in some cases)
most of old code uses constants and so does not handle things right.

instead you should be fetching tile size from the tensor spec.

jbbieniekTT replied:
Done

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/scatter/scatter.hpp:
```diff
@@ -0,0 +1,36 @@
+
+#pragma once
+
+#include "scatter_enums.hpp"
```
using `ScatterReductionType` from this include makes sense, but this include also leaks lower level things. I am not a fan of this

jbbieniekTT replied:
Moved to `scatter_program_factory.hpp`

cmaryanTT commented on tests/ttnn/unit_tests/operations/test_scatter.py:
```diff
@@ -0,0 +1,70 @@
+        ([2, 30, 200], -1, [2, 30, 200], ttnn.float32),
+        ([1, 1, 20, 20, 200], -1, [1, 1, 20, 20, 20], ttnn.bfloat16),
+        ([2, 2, 2, 2, 2, 2, 2, 2], -1, [2, 2, 2, 2, 2, 2, 2, 2], ttnn.float32),
+        ([10, 1, 10, 1, 10], -1, [10, 1, 10, 1, 1], ttnn.bfloat16),
```
Please include large shapes, including the client shapes. Part of the claim of this PR is support for tensors of any size.

jbbieniekTT replied:
@cmaryanTT @bbradelTT @aczajkowskiTT i have managed to reduce execution time to under one minute for BH, i am including this case here

philei-tt replied:
@jbbieniekTT  Here: https://github.com/tenstorrent/tt-metal/pull/22731 I will add @pytest.mark.slow marker, so that slow tests are not ran by default, and you can run them with --runslow 

Maybe that will be useful for you

cmaryanTT replied:
I am ok with the large shape being left out or marked as slow as Pavlo suggested. Next improvement to the OP should be focused on large shape support.

jbbieniekTT replied:
@philei-tt that's excellent, i am very happy to mark my test with this as soon as the PR gets merged, if not in this PR, then in the PR that will follow immediately after, the one focusing on optimising unfortunate input shapes
@bbradelTT i will as well create a sweep test in this PR, i think
@cmaryanTT @aczajkowskiTT i am going to refactor the last thing from @ayerofieiev-tt today and merge the PR asap, the issue has been created (#22926)

ayerofieiev-tt reviewed (commented):
Current tests might be insufficient and I recommend to extend tests to include:
* All dimensions (dim=0,1,2,-1,-2,-3)
* Edge cases (empty tensors, boundary indices)
* Error conditions (invalid dimensions, out-of-bounds indices)
* Data Types (integer, bfloat8_b)

jbbieniekTT committed:
16942: address PR comments

jbbieniekTT commented on ttnn/cpp/ttnn/operations/experimental/scatter/device/scatter_device_operation.cpp:
```diff
@@ -0,0 +1,177 @@
+            static_cast<int32_t>(dim) < static_cast<int32_t>(output_rank),
+            "dim must be lower than output shape's positive rank (dim: {}, rank: {}).",
+            dim,
+            static_cast<int32_t>(output_rank));
```
Done

jbbieniekTT commented on ttnn/cpp/ttnn/operations/experimental/scatter/device/kernels/dataflow/reader_scatter.cpp:
```diff
@@ -0,0 +1,169 @@
+                            // check if index value targets currently chosen input tile (tile_input)
+                            const uint32_t dest_tile_id_in_row = index_value >> 5;
+                            if (dest_tile_id_in_row != tile_input) {
+                                ASSERT(dest_tile_id_in_row == tile_input);
```
Fixed the assert itself als

jbbieniekTT committed:
16942: address PR comments

rfurko-tt reviewed (changes requested):
```
const Tensor transposed_tensor = reduction_common::perform_transpose(input_tensor, is_dim_last_idx, dim, -1);
```

i really don't like this approach, we should use transpose really careful with TILED layout. 

Imagine, that you need to perform transpose on tensor with shape [1, 2, N, M] (-3 and -1), now shape of the new tensor is [1, M, N, 2], but actually it is [1, M, N, 2[32]] as last dimension will be tiled. As you can see, now tensor volume increased 16x. I don't think it is expected behavior of the scatter operation and potentially a problem which will be hard to debug/understand.

rfurko-tt commented:
"this approach enables easy and cache-friendly implementation"

* I agree that it simplifies implementation of the operation, but it does not cover use-cases when it should work as well
* "cache-friendly implementation" program cache? are we optimizing by this criteria? 

"the size may either increase or decrease (depending whether dim1 > dim2 or dim1 < dim2 between 1x and 32x times"
the problem when it increases 32x times

aczajkowskiTT commented:
> ```
> const Tensor transposed_tensor = reduction_common::perform_transpose(input_tensor, is_dim_last_idx, dim, -1);
> ```
> 
> i really don't like this approach, we should use transpose really careful with TILED layout.
> 
> Imagine, that you need to perform transpose on tensor with shape [1, 2, N, M] (-3 and -1), now shape of the new tensor is [1, M, N, 2], but actually it is [1, M, N, 2[32]] as last dimension will be tiled. As you can see, now tensor volume increased 16x. I don't think it is expected behavior of the scatter operation and potentially a problem which will be hard to debug/understand.

@rfurko-tt I don't think it's wise from a software engineering point of view to be planning around a specific OP's bugs and implementation errors.  Transpose is the correct operation to use in this case no?  If there is a corner case that transpose cannot handle in L1, it should have an exit case to DRAM, or should be fixed; rather than our unrelated OP keeping track of transpose bugs.  Am I wrong on this?

On another note, for the customer shape we are interested in, I understand the transpose issue does not occur.

aczajkowskiTT commented:
> @aczajkowskiTT
> 
> "I don't think it's wise from a software engineering point of view to be planning around a specific OP's bugs and implementation errors." - it's not specific bug in the transpose operation... that's how tiling works
> 
> Imagine two tensors: tensor A: [1, 2, 16384, 16384] -> fits in DRAM (way less < 12 GB) tensor B: [16384, 2, 16384, 1] -> does not fit in DRAM (more than 12 GB)
> 
> please review discussion in the `ttnn::sum` issue i've shared

Right so based on a specific input shape into transpose, wouldn't it be cleaner from a usability and developer perspective for transpose to handle that corner case?  Then if transpose returns an error code you default to another approach (your preferred alternative)?

aczajkowskiTT commented:
> "wouldn't it be cleaner from a usability and developer perspective for transpose to handle that corner case?"
> 
> with all due respect, it's not issue of transpose. It is expected that transpose will use that much memory. you are creating tensor that padded shape requires 17 GB. The issue is that scatter calls transpose.
> 
> transpose will return error code, to be precise - throw exception "tensor does not fit in DRAM"

Okay, so catch the error code, and use a different approach, L1 cache/number of cores will change with architectures, so shapes that transpose can or cannot handle is hardware dependent, thus it is below the abstraction layer of what the op does.

aczajkowskiTT commented:
> "Regarding exception, I think our current version of Python should support zero-cost exceptions, so there is no cost if no exception is raised.
> 
> you are trying to prove why it's okay to release half-baked operation... which is not acceptable for providing to our customers.
> 
> i would never use hardware which require me to write code like this
> 
> ```
>     try:
>       f0 = x + y
>     catch:
>       raise Exception("Unfortunately our operation takes more memory than needed")
> 
>     try:
>       f1 = y + x
>     catch:
>       raise Exception("Unfortunately our operation takes more memory than needed")
> ```
> 
> "@rfurko-tt I think we're having a philosophical discussion at this point, which is where the disagreement is. I think optimizing for perf with more PRs is totally valid, as you suggest."
> 
> this is not philosophical discussion, I am trying to show you that operation has unexpected behavior which results in 32x more memory consumption and it's not an issue of any other operation other than this one.

God forbid we throw and catch exceptions...

aczajkowskiTT commented:
@rfurko-tt @ayerofieiev-tt @jbbieniekTT After offline discussion, we decided that for problematic tensor shapes that can cause a significant increase in memory usage, this implementation will need to be amended with another PR to detect these shapes and handle them explicitly.

@jbbieniekTT Please create a ticket against this PR for the shapes @rfurko-tt is concerned about.  Include performance profiling as part of that.

cmaryanTT reviewed (approved):
Approving to get an MVP of this OP committed. Performant support for larger shapes can follow in the next PR, which should get a new issue opened; tests for the larger shape of interest can be either in this PR and marked as slow, or left out for the next PR. Longer term followup changes to improve behaviour around or remove transpose should be tracked in a separate issue.

dmakoviichuk-tt commented:
HI @cmaryanTT and @jbbieniekTT  I suggest to throw exception for the case where transpose would work 10x slower and/or use much more memory:
When we are transposing and one of two last dimensions is not divisible by TILE.  
And please create ticket  with high priority to support.
We saw similar decision with other ops. As result we wasted couple of days debugging the issue because we could not expect **ttnn::sum** can be so slow. Anyway noone is going to use this op for these cases. It might throw exception or just be too slow. In our case ttnn::sum with the same issue was as slow as a half of gpt2s :)
I don't thin we need this "fake" generality. It might save a lot time for the customers.

I will approve this pr immediately once I see:
1) a link to the created ticket.
2) Exception for these really bad cases to save customers time.
3) updated pr description for the op behavior.
<img width="519" alt="Screenshot 2025-06-02 at 7 40 36 PM" src="https://github.com/user-attachments/assets/b49f0cb5-87be-4961-a569-cfd036510b89" />
According to the description it should work with any tensor sizes which is not true.
It should not take a lot of time to do it
Thanks

jbbieniekTT committed:
16942: ignore padding better

ignore padding in a more optimized way

jbbieniekTT commented on ttnn/cpp/ttnn/operations/experimental/scatter/scatter.hpp:
```diff
@@ -0,0 +1,36 @@
+        const std::optional<MemoryConfig>& opt_out_memory_config,
+        const std::optional<scatter::ScatterReductionType>& opt_reduction,
+        std::optional<Tensor>& opt_output,
+        const QueueId& queue_id = DefaultQueueId);
```
Okay, it's fixed

jbbieniekTT committed:
16942: address PR comments

jbbieniekTT committed:
16942: implement constraints

implement restrictions as discussed

jbbieniekTT commented:
> @dmakoviichuk-tt that's reasonable. @jbbieniekTT - can you please add a TT_FATAL for whatever size approximately exceeds ~5 seconds of execution time (unusably long, but not so long that it's annoying for isolated usage), and another for
> 
> * for last two dims not divisible by tile size
> * AND we are transposing
> * AND the transposed tensor size would be > 200KB larger than its original size (i.e. for small tensors, the transpose is fine, because the absolute increase is reasonable)

Done

rfurko-tt reviewed (approved):
LGTM, thanks for the changes

Please add ticket to fix program cache as discussed offline. thanks in advance
